tests/unit/test_probes_provider.py

```diff
@@ -0,0 +1,195 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import json
+import unittest
+from typing import List
+
+from charms.blackbox_k8s.v0.blackbox_probes import BlackboxProbesProvider
+from cosl import JujuTopology
+from ops.charm import CharmBase
+from ops.framework import StoredState
+from ops.model import (
+    ActiveStatus,
+    BlockedStatus,
+)
+from ops.testing import Harness
+
+RELATION_NAME = "probes"
+
+PROVIDER_META = f"""
+name: provider-tester
+containers:
+  blackbox-tester:
+provides:
+  {RELATION_NAME}:
+    interface: blackbox_exporter_probes
+"""
+
+
+PROBES: List[dict] = [
+    {
+        "job_name": "my-first-job",
+        "params": {"module": ["http_2xx"]},
+        "static_configs": [
+            {
+                "targets": ["10.1.238.1"],
+                "labels": {"some_key": "some-value"},
+            }
+        ],
+    },
+    {
+        "job_name": "my-second-job",
+        "params": {
+            "module": ["icmp"],
+        },
+        "static_configs": [
+            {"targets": ["10.1.238.1"], "labels": {"some_other_key": "some-other-value"}}
+        ],
+    },
+]
+
+PROBES_NOT_VALID_MISSING_STATIC_CONFIG: List[dict] = [
+    {
+        "job_name": "my-first-job",
+        "params": {"module": ["http_2xx"]},
+    }
+]
+
+PROBES_NOT_VALID_MISSING_MODULE: List[dict] = [
+    {
+        "job_name": "my-first-job",
+        "static_configs": [
+            {
+                "targets": ["10.1.238.1"],
+                "labels": {"some_key": "some-value"},
+            }
+        ],
+    },
+]
+
+MODULES: dict = {
+    "http_2xx_longer_timeout": {
+        "prober": "http",
+        "timeout": "30s",
+    }
+}
+
+
+class BlackboxProbesProviderCharmWithModules(CharmBase):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args)
+
+        self.provider = BlackboxProbesProvider(self, probes=PROBES, modules=MODULES)
+
+
+class BlackboxProbesProviderTest(unittest.TestCase):
+    def setUp(self):
+        self.harness = Harness(BlackboxProbesProviderCharmWithModules, meta=PROVIDER_META)
+        self.harness.set_model_name("MyUUID")
+        self.addCleanup(self.harness.cleanup)
+        self.harness.set_leader(True)
+        self.harness.begin()
```

Comment:
Any reason not to use `begin_with_initial_hooks` here? I find it to be more useful, i.e. giving more confidence.

Reply:
changed in 8eaaac8170f5fdc10c1d2483b154b262fa145ead